front/src/app/services/wallet.service.ts

```diff
@@ -17,30 +17,116 @@ import {Transaction} from '../models/transaction.model';
 /*UTILS*/
 import {objIsEmpty} from '../utils/functions';
 
+interface IWallet {
+  w3: Web3;
+
+  logged$: BehaviorSubject<boolean>;
+
+  send(tx: TransactionConfig): Observable<TransactionReceipt>;
+
+  call(tx: TransactionConfig): Observable<string>;
+
+  logIn(privateKey: string): Observable<string>;
+}
+
+abstract class Wallet {
+  w3: Web3;
+  logged$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(null);
+
+  constructor(w3: Web3) {
+    this.w3 = w3;
+  }
+}
+
+class MetamaskStrategy extends Wallet implements IWallet {
+
+  send(tx: TransactionConfig): Observable<TransactionReceipt> {
+    return fromPromise(this.w3.eth.sendTransaction(tx));
+  }
+
+  call(tx: TransactionConfig): Observable<string> {
+    return fromPromise(this.w3.eth.call(tx));
+  }
+
+  logIn(): Observable<string> {
+    try {
+      (window as any).ethereum.enable();
+      return fromPromise(this.w3.eth.getAccounts()).pipe(
+        map((v: string[]) => v[0]),
+      );
+    } catch (e) {
+      return throwError('Access not granted');
+    }
+  }
+}
+
+class PrivateKeyStrategy extends Wallet implements IWallet {
+  private privateKey: string;
+  private account: Account;
+
+  send(tx: TransactionConfig): Observable<TransactionReceipt> {
+    return fromPromise(this.w3.eth.signTransaction(tx)).pipe(
+      flatMap((signedTx: RLPEncodedTransaction) => fromPromise(this.w3.eth.sendSignedTransaction(signedTx.raw))),
+    );
+  }
+
+  call(tx: TransactionConfig): Observable<string> {
+    return fromPromise(this.w3.eth.call(tx));
+  }
+
+  logIn(privateKey: string): Observable<string> {
+    if (privateKey.length === 64 && privateKey.indexOf('0x') !== 0) {
+      privateKey = '0x' + privateKey;
+    }
+    if (privateKey.length === 66) {
+      this.account = this.w3.eth.accounts.privateKeyToAccount(privateKey);
+      this.logged$.next(true);
+      return of(this.account.address);
+    }
+    return throwError('Given private key is not valid');
+  }
+}
+
 @Injectable()
 export class WalletService {
   isProcessing = false;
 
   // ACCOUNT INFO
   account: Account;
+  accountAddress: string;
+  accountAddress$: BehaviorSubject<string> = new BehaviorSubject<string>(null);
   accountBalance: string;
 
   receipt: TransactionReceipt;
 
+  metamaskIntalled$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  metamaskConfigured$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  logged$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+
+  private _ready$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  get ready$(): Observable<boolean> {
+    return this._ready$.pipe(
+      filter<boolean>(v => !!v),
+      take(1),
+    );
+  }
+
   private _web3Callable$: BehaviorSubject<Web3> = new BehaviorSubject(null);
-  private _web3Payable$: BehaviorSubject<Web3> = new BehaviorSubject(null);
+  private _web3Metamask$: BehaviorSubject<Web3> = new BehaviorSubject(null);
```

Comment:
Switching back to Metamask specific naming seems backwards. Why don't we keep these general, for example:
```
walletInstalled$:
walletConfigured$:
_web3Payable$:
```
Then if/when we switch or add support for others, none of this needs to be updated. 
(The user messages can stay metamask specific until we actually add more)

Reply:
@jmank88 can you check new code, wallet and metamask is different, pk is also wallet